docs/snapshotting/snapshot-support.md

```diff
@@ -0,0 +1,280 @@
+[MAP_PRIVATE mapping](http://man7.org/linux/man-pages/man2/mmap.2.html) of the
+memory file, resulting in runtime on-demand loading of memory pages. Any subsequent
+memory writes go to a copy-on-write anonymous memory mapping.
+This has the advantage of very fast snapshot loading times, but comes with the cost
+of having to keep the guest memory file around for the entire lifetime of the
+resumed microVM.
+
+## Firecracker Snapshotting characteristics
+
+- Fresh Firecracker microVMs are booted using `anonymous` memory, while microVMs
+  resumed from snapshot load memory on-demand from the snapshot and copy-on-write
+  to anonymous memory.
+- Resuming from a snapshot is optimized for speed, while taking a snapshot involves
+  some extra CPU cycles for synchronously writing dirty memory pages to the memory
+  snapshot file. Taking a snapshot of a fresh microVM, on which dirty pages tracking
+  is not enabled, results in full contents of guest memory being written to the
+  snapshot.
+- The _memory file_ and _microVM state file_ are generated by Firecracker on snapshot
+  creation and the disk contents are flushed to their backing files.  
+- The API calls exposing the snapshotting functionality have clear **Prerequisites**
+  that describe the requirements on when/how they should be used.
+ 
+## Snapshot versioning
+
+Firecracker snapshotting implementation offers support for microVM versioning
+(`cross-version snapshots`) in the following contexts:
+- saving snapshots at older versions (being able to create a snapshot with any version
+in `[N, N + o]` interval while being in Firecracker version `N+o`),
+- loading snapshots from older versions (being able to load any snapshot created by
+Firecracker version `N` in a Firecracker version `N+o`).
+
+The design supports an unlimited number of versions, the value of `o` (maximum number
+of older versions that we can restore from / save a snapshot to, from the current
+version) will be defined later.
+
+## Snapshot API
+
+Firecracker exposes the following APIs for manipulating snapshots: `Pause`, `Resume`
+and `CreateSnapshot` can be called only after booting the microVM, while `LoadSnapshot`
+is allowed only before boot.
+
+### Pausing the microVM
+
+To create a snapshot, you have to firstly pause the running microVM and its vCPUs with
+the following API command:
+
+```bash
+curl --unix-socket /tmp/firecracker.socket -i \
+    -X PATCH "http://localhost/vm" \
+    -H "accept: application/json" \
+    -H "Content-Type: application/json" \
+    -d "{
+            "state": "Paused"
+    }"
+```
+
+**Prerequisites**: Pausing the microVM can be done any time after the microVM is booted.
+                   Successive calls of this request keep the microVM in the `Paused`
+                   state.
+**Effects**:
+- _on success_: microVM is guaranteed to be `Paused`.
+- _on failure_: no side-effects.
+
+*Note*: If you change your mind and want to create a snapshot later, you can resume the
```

Comment:
I would remove this note. This can be rephrased better in the context of creating snapshots, highlighting the possible flows of boot -> pause -> save -> resume -> pause -> save, load->resume->pause->save->resume which work with either diff or full snapshots.

Reply:
Done, I updated with these flows at the end of the doc (after the `load` part).